### ISSUE TITLE
minio gateway s3: infinite loop in cleanupStaleUploads of encryption multipart uploads

### ISSUE THREAD

**Issue by catac** [minio/minio#10588]
## Expected Behavior
When KMS encryption is configured in the s3 minio gateway, the cleanup of encryption multipart uploads should work correctly.

## Current Behavior
When KMS encryption is configured on minio gateway, a timer is created to [cleanup incomplete encrypted](https://github.com/minio/minio/blob/80fab03b630943118211fa04a02409da32dd3f4b/cmd/gateway/s3/gateway-s3-sse.go#L695) multipart uploads (each 24h). The task `cleanupStaleUploads()` runs the cleanup in a loop without a proper exit condition. It finishes only if the configured user cannot lists its own s3 buckets, otherwise it loops continuously.

## Possible Solution
The [cleanupStaleUploads()](https://github.com/minio/minio/blob/80fab03b630943118211fa04a02409da32dd3f4b/cmd/gateway/s3/gateway-s3-sse.go#L711-L722) function should be changed so that:
- error is logged if the `ListBuckets` call fails
- surrounding `for { ... }` could probably be removed or some exit condition should be present so that, if `ListBuckets` doesn't fail, the function exits after processing the objects.
- while here, I think the error of the [ListObjectsV2()](https://github.com/minio/minio/blob/80fab03b630943118211fa04a02409da32dd3f4b/cmd/gateway/s3/gateway-s3-sse.go#L730) call in `getStalePartsForBucket` should be logged as well. I'm not sure if the processing continues if some prefixes are not allowed.

## Steps to Reproduce (for bugs)

1. start minio with KMS encryption, for example
```
export MINIO_ACCESS_KEY=xxxxx
export MINIO_SECRET_KEY=yyyyy
export MINIO_KMS_AUTO_ENCRYPTION=on
export MINIO_KMS_MASTER_KEY=zzzzz
minio gateway s3 
```
2. the given ACCESS/SECRET keys should have the permission to list own buckets
3. wait 24h (or change the [GlobalStaleUploadsCleanupInterval](https://github.com/minio/minio/blob/80fab03b630943118211fa04a02409da32dd3f4b/cmd/globals.go#L85) to less than that)
3. the minio process will consume ~10% CPU and perform lots of queries to S3
4. you can get a goroutines dump via `kill -SIGABRT <pid>`:
```
goroutine 65 [select]:
net/http.(*persistConn).roundTrip(0xc0014490e0, 0xc0013704c0, 0x0, 0x0, 0x0)
net/http/transport.go:2559 +0x763
net/http.(*Transport).roundTrip(0xc001052500, 0xc0000f2f00, 0x150, 0x150, 0xc00008d8c0)
net/http/transport.go:582 +0xa88
net/http.(*Transport).RoundTrip(...)
net/http/roundtrip.go:17
github.com/minio/minio/cmd.MetricsTransport.RoundTrip(0xc001052500, 0xc000efcc00, 0xc0000f2f00, 0xc001370480, 0xc0012d2338, 0x40d1a8)
github.com/minio/minio/cmd/gateway-common.go:411 +0xb6
net/http.send(0xc0000f2f00, 0x2103880, 0xc001013850, 0x0, 0x0, 0x0, 0xc0004da020, 0x1d20a99, 0x1, 0x0)
net/http/client.go:252 +0x43e
net/http.(*Client).send(0xc0010338f0, 0xc0000f2f00, 0x0, 0x0, 0x0, 0xc0004da020, 0x0, 0x1, 0x0)
net/http/client.go:176 +0xfa
net/http.(*Client).do(0xc0010338f0, 0xc0000f2f00, 0x0, 0x0, 0x0)
net/http/client.go:699 +0x44a
net/http.(*Client).Do(...)
net/http/client.go:567
github.com/minio/minio-go/v7.Client.do(0xc000f39000, 0xc001007980, 0x0, 0x0, 0x0, 0x0, 0x0, 0x1, 0xc0010338f0, 0xc001050360, ...)
github.com/minio/minio-go/v7@v7.0.5-0.20200811211821-14ed05478889/api.go:472 +0x50
github.com/minio/minio-go/v7.Client.executeMethod(0xc000f39000, 0xc001007980, 0x0, 0x0, 0x0, 0x0, 0x0, 0x1, 0xc0010338f0, 0xc001050360, ...)
github.com/minio/minio-go/v7@v7.0.5-0.20200811211821-14ed05478889/api.go:575 +0x38a
github.com/minio/minio-go/v7.Client.listObjectsV2Query(0xc000f39000, 0xc001007980, 0x0, 0x0, 0x0, 0x0, 0x0, 0x1, 0xc0010338f0, 0xc001050360, ...)
github.com/minio/minio-go/v7@v7.0.5-0.20200811211821-14ed05478889/api-list.go:201 +0x49c
github.com/minio/minio-go/v7.Core.ListObjectsV2(0xc000e58580, 0xc000bac724, 0x7, 0x0, 0x0, 0xc000194c40, 0x6a, 0x203000, 0x0, 0x0, ...)
github.com/minio/minio-go/v7@v7.0.5-0.20200811211821-14ed05478889/core.go:55 +0x17a
github.com/minio/minio/cmd/gateway/s3.(*s3Objects).ListObjectsV2(0xc000994480, 0x213cae0, 0xc000328fc0, 0xc000bac724, 0x7, 0x0, 0x0, 0xc000194c40, 0x6a, 0x0, ...)
github.com/minio/minio/cmd/gateway/s3/gateway-s3.go:378 +0xf8
github.com/minio/minio/cmd/gateway/s3.(*s3EncObjects).getStalePartsForBucket(0xc000994480, 0x213cae0, 0xc000328fc0, 0xc000bac724, 0x7, 0x4e94914f0000, 0x0)
github.com/minio/minio/cmd/gateway/s3/gateway-s3-sse.go:730 +0x10b
github.com/minio/minio/cmd/gateway/s3.(*s3EncObjects).cleanupStaleUploads(0xc000994480, 0x213cae0, 0xc000328fc0, 0x4e94914f0000)
github.com/minio/minio/cmd/gateway/s3/gateway-s3-sse.go:717 +0x20f
github.com/minio/minio/cmd/gateway/s3.(*s3EncObjects).cleanupStaleEncMultipartUploads(0xc000994480, 0x213cae0, 0xc000328fc0, 0x4e94914f0000, 0x4e94914f0000)
github.com/minio/minio/cmd/gateway/s3/gateway-s3-sse.go:704 +0xb0
created by github.com/minio/minio/cmd/gateway/s3.(*S3).NewGatewayLayer
github.com/minio/minio/cmd/gateway/s3/gateway-s3.go:244 +0x3aa
```

## Context
I am trying to encrypt data sent to S3 as close as possible to the source (SSE is not enough and SSE-C is not supported by other tools). Minio gateway seems like a great solution for this.

## Regression
No, this part of the code seems unchanged since long time ago.

## Your Environment

* Version used (`minio --version`): `minio version RELEASE.2020-09-23T19-18-30Z`
* Server setup and configuration: linux minio binary, with a configuration similar to the one given in the "Steps to Reproduce" section
* Operating System and version (`uname -a`): Flatcar Linux 2512.4.0

**Comment by harshavardhana** [comment]
Fixed by #10840

**Comment by StevenReitsma** [comment]
I don't think this is actually fixed with #10840, but #10845 should
